src/lib.rs

```diff
@@ -18,6 +18,27 @@
 #[cfg(feature = "alloc")]
 extern crate alloc;
 
+// This must be in the root lib.rs, otherwise it doesn't appear in the final binary.
+
+#[used]
+#[cfg_attr(target_os = "macos", link_section = ".defmt,end.VERSION")]
+#[cfg_attr(not(target_os = "macos"), link_section = ".defmt.end")]
+#[export_name = "_defmt_version_ = 3"]
+static DEFMT_VERSION: u8 = 0;
```

Comment:
It would be good to have some docs on this.

```suggestion
/// The defmt ABI and wire format version.
///
/// This number has to be updated every time there is a backwards-incompatible change to
/// - the symbol naming scheme
/// - the symbol and section layout
/// - the data encoding / wire format
#[used]
#[cfg_attr(target_os = "macos", link_section = ".defmt,end.VERSION")]
#[cfg_attr(not(target_os = "macos"), link_section = ".defmt.end")]
#[export_name = "_defmt_version_ = 3"]
static DEFMT_VERSION: u8 = 0;
```

Reply:
Is there a way to catch these backwards-incompatible changes in the CI?

Reply:
We can probably come up with a backwards-compat check again, yes